p2p/shim.go

```diff
@@ -0,0 +1,227 @@
+package p2p
+
+import (
+	"fmt"
+
+	"github.com/gogo/protobuf/proto"
+)
+
+var _ Reactor = (*ReactorShim)(nil)
+
+type (
+	CodecShim interface {
+		Marshal(proto.Message) ([]byte, error)
+		Unmarshal([]byte) (proto.Message, error)
+		Validate(proto.Message) error
+	}
+
+	ChannelShim struct {
+		Descriptor *ChannelDescriptor
+		Channel    *Channel
+		InCh       chan Envelope
+		OutCh      chan Envelope
+		PeerErrCh  chan PeerError
+	}
+
+	ReactorShim struct {
+		BaseReactor
+
+		Name         string
+		Codec        CodecShim
+		PeerUpdateCh chan PeerUpdate
+		Channels     map[ChannelID]*ChannelShim
+	}
+)
+
+func NewShim(name string, codec CodecShim, impl Reactor, descriptors []*ChannelDescriptor) *ReactorShim {
+	br := *NewBaseReactor(name, impl)
+
+	channels := make(map[ChannelID]*ChannelShim)
+	for _, cd := range descriptors {
+		cID := ChannelID(cd.ID)
+		inCh := make(chan Envelope)
+		outCh := make(chan Envelope)
+		peerErrCh := make(chan PeerError)
+
+		channels[cID] = &ChannelShim{
+			Descriptor: cd,
+			Channel:    NewChannel(cID, nil, inCh, outCh, peerErrCh),
+			InCh:       inCh,
+			OutCh:      outCh,
+			PeerErrCh:  peerErrCh,
+		}
+	}
+
+	return &ReactorShim{
+		BaseReactor:  br,
+		Name:         name,
+		Codec:        codec,
+		PeerUpdateCh: make(chan PeerUpdate),
+		Channels:     channels,
+	}
+}
+
+// proxyPeerErrors iterates over each p2p Channel and starts a separate go-routine
+// where we listen for peer errors on the channel's PeerErrCh and send a PeerUpdate
+// on the PeerUpdateCh with a status PeerStatusBanned.
```

Comment:
Why do we have to do this? PeerUpdates are updates from the switch to the reactor, while errors are from the reactor to the switch. If the peer reports an error, it already knows about it, so shouldn't get an update.

If we receive a peer error, we'll have to stop the peer if requested by the reactor (StopPeerForError or something), which I think should trigger a `RemovePeer` call. The exact error severities and semantics aren't really clear yet, but I suggest we temporarily just pass explicit actions since that's what the current code assumes (e.g. `PeerErrorSeverityStop` and `PeerErrorSeverityIgnore`) - we can come up with better severities when we design the router.

Reply:
Correct, this is the understanding I have already. The rationale I had for doing this, and now I see it's completely unnecessary, was that this was kind of a temporary filler to act as the new router (until we have it).

In other words, reactors would send `PeerError` messages on the respective channel and the shim would pick it up (in `proxyPeerErrors`) and proxy it to `PeerUpdateCh` kind of acting as temporary router. 

But yeah, TL;DR I see that `proxyPeerErrors` is unnecessary.